pkg/cloudprovider/aws/aws.go

```diff
@@ -489,3 +808,338 @@ func (self *AWSCloud) GetZone() (cloudprovider.Zone, error) {
+	}
+
+	group := groups[0]
+
+	for _, permission := range group.IPPerms {
+		if permission.FromPort != fromPort {
+			continue
+		}
+		if permission.ToPort != toPort {
+			continue
+		}
+		if permission.Protocol != protocol {
+			continue
+		}
+		if len(permission.SourceIPs) != 1 {
+			continue
+		}
+		if permission.SourceIPs[0] != sourceIp {
+			continue
+		}
+		return false, nil
+	}
+
+	newPermission := ec2.IPPerm{}
+	newPermission.FromPort = fromPort
+	newPermission.ToPort = toPort
+	newPermission.SourceIPs = []string{sourceIp}
+	newPermission.Protocol = protocol
+
+	newPermissions := []ec2.IPPerm{newPermission}
+	_, err = self.ec2.AuthorizeSecurityGroupIngress(securityGroupId, newPermissions)
+	if err != nil {
+		glog.Warning("error authorizing security group ingress", err)
+		return false, err
+	}
+
+	return true, nil
+}
+
+// CreateTCPLoadBalancer implements TCPLoadBalancer.CreateTCPLoadBalancer
+func (self *AWSCloud) CreateTCPLoadBalancer(name, region string, externalIP net.IP, port int, hosts []string, affinity api.AffinityType) (string, error) {
+	glog.V(2).Infof("CreateTCPLoadBalancer(%v, %v, %v, %v, %v)", name, region, externalIP, port, hosts)
+
+	if affinity != api.AffinityTypeNone {
+		// ELB supports sticky sessions, but only when configured for HTTP/HTTPS
+		return "", fmt.Errorf("unsupported load balancer affinity: %v", affinity)
+	}
+
+	if len(externalIP) > 0 {
+		return "", fmt.Errorf("External IP cannot be specified for AWS ELB")
+	}
+
+	instances, err := self.getInstancesByDnsNames(hosts)
+	if err != nil {
+		return "", err
+	}
+
+	vpc, err := self.findVpc()
+	if err != nil {
+		glog.Error("error finding vpc", err)
+		return "", err
+	}
+	if vpc == nil {
+		return "", fmt.Errorf("Unable to find vpc")
+	}
+
+	// Construct list of configured subnets
+	subnetIds := []string{}
+	{
+		subnets, err := self.ec2.DescribeSubnets(nil, vpc.VpcId)
+		if err != nil {
+			return "", err
+		}
+
+		//	zones := []string{}
+		for _, subnet := range subnets {
+			subnetIds = append(subnetIds, subnet.SubnetId)
+			if !strings.HasPrefix(subnet.AvailabilityZone, region) {
+				glog.Error("found AZ that did not match region", subnet.AvailabilityZone, " vs ", region)
+				return "", fmt.Errorf("invalid AZ for region")
+			}
+			//		zones = append(zones, subnet.AvailabilityZone)
+		}
+	}
+
+	// Build the load balancer itself
+	var loadBalancerName, dnsName string
+	{
+		loadBalancer, err := self.describeLoadBalancer(region, name)
+		if err != nil {
+			return "", err
+		}
+
+		if loadBalancer == nil {
+			createRequest := &elb.CreateLoadBalancer{}
+			// TODO: Is there a k8s UUID that it would make sense to use?
+			uuid := strings.Replace(string(util.NewUUID()), "-", "", -1)
+			awsId := LOADBALANCER_NAME_PREFIX + uuid
+			if len(awsId) > LOADBALANCER_NAME_MAXLEN {
+				awsId = awsId[:LOADBALANCER_NAME_MAXLEN]
+			}
+			createRequest.LoadBalancerName = awsId
+
+			listener := elb.Listener{}
+			listener.InstancePort = int64(port)
```

Comment:
I haven't thought this through entirely, but if the nodes' public IPs are the ones being targeted by the load balancer, does using the service's port prevent multiple services from listening on the same port?